File: 01-CS-Basic/02-数据结构与算法/Swift/HZDataStructure/09-Tree/Trie.swift
//
//  Trie.swift
//  HZDataStructure
//
//  Created by 华天杰 on 2021/1/4.
//

import Cocoa

/*
  实现前缀树：
 
 */
class Trie: NSObject {
    
    class Node{
        var isWord: Bool!
        var next: [Character:Node]!
//        // 指定构造器
//        init(isWord:Bool) {
//            self.isWord = isWord
//            self.next = [Character:Node]()
//        }
        // 遍历构造器
         init(){
            self.isWord = false
            self.next = [Character:Node]()
        }
    }
    
    private var root:Node!
    private var size:Int!
    
    override init() {
        self.root =  Node()
        self.size = 0
    }
    
    // 向Tire中添加单词
    public func insert(_ word:String){
        var cur = root!
        for char in word {
            if cur.next[char] == nil {
                cur.next[char] = Node()
            }
            cur = cur.next[char]!
        }
        if !cur.isWord {
            cur.isWord = true
            self.size += 1
        }
    }
    // 在Tire中查询是否存在指定的word
    public func search(_ word:String) -> Bool{
        var cur = root!
        for char in word {
            if cur.next[char] == nil {
                return false
            }
            cur = cur.next[char]!
        }
        return cur.isWord
    }
    // 在Tire中搜索以prefix为开头的单词是否存在
    public func startsWith(_ prefix:String) ->Bool{
        var cur = root!
        for char in prefix {
            if cur.next[char] == nil {
                return false
            }
            cur = cur.next[char]!
        }
        return true
    }
    
}

File: 01-CS-Basic/02-数据结构与算法/Java/DS/src/Trie/Trie.java
package Trie;

import java.util.TreeMap;

public class Trie {

    private class Node{

        public boolean isWord;
        public TreeMap<Character,Node> next;

        public Node(boolean isWord){
            this.isWord = isWord;
            next = new TreeMap<>();
        }

        public Node(){
            this(false);
        }
    }

    private Node root;
    private int size;

    public Trie(){
        root = new Node();
        size = 0;
    }

    // 获取Trie中存储的单词数量
    public int getSize(){
        return size;
    }

    // 向Trie中添加一个新的单词word
    public void insert(String word){
        Node cur = root;
        for (int i = 0; i < word.length() ; i++) {
            char c = word.charAt(i);
            if (cur.next.get(c) == null){
                cur.next.put(c,new Node());
            }
            cur = cur.next.get(c);
        }
        if (!cur.isWord){
           cur.isWord = true;
           size ++;
        }
    }

   // 查询单词word是否在Trie中
   public boolean search(String word){

        Node cur = root;
       for (int i = 0; i < word.length(); i++) {
           char c = word.charAt(i);
           if (cur.next.get(c) == null){
               return false;
           }
           cur = cur.next.get(c);
       }
       return  cur.isWord; // 避免遍历都结尾不是一个单词的情况
   }
   
   // 查询是否在Trie中有单词以prefix为前缀
    public boolean startsWith(String prefix){

        Node cur = root;
        for (int i = 0; i < prefix.length(); i++) {
            char c = prefix.charAt(i);
            if (cur.next.get(c) == null){
                return false;
            }
            cur = cur.next.get(c);
        }
        return true;
    }
}
